src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/CryptographicAttributeObjectCollection.cs

```diff
@@ -0,0 +1,158 @@
+        }
+
+        public int Add(CryptographicAttributeObject attribute)
+        {
+            if (attribute == null)
+                throw new ArgumentNullException(nameof(attribute));
+
+            //
+            // Merge with existing attribute, if already existed, else add as new.
+            //
+            string szOid1 = attribute.Oid.Value;
+            for (int index = 0; index < _list.Count; index++)
+            {
+                CryptographicAttributeObject existing = _list[index];
+
+                // To prevent caller to add the existing item into the collection again
+                // Otherwise the merge will be an infinite loop
+                if (object.ReferenceEquals(existing.Values, attribute.Values))
+                    throw new InvalidOperationException(SR.InvalidOperation_DuplicateItemNotAllowed);
+
+                string szOid2 = existing.Oid.Value;
+                if (string.Equals(szOid1, szOid2, StringComparison.OrdinalIgnoreCase))
+                {
+                    //
+                    // Only allow one signing time, per RFC.
+                    //
+                    if (string.Equals(szOid1, Oids.SigningTime, StringComparison.OrdinalIgnoreCase))
+                        throw new CryptographicException(SR.Cryptography_Pkcs9_MultipleSigningTimeNotAllowed);
+
+                    foreach (AsnEncodedData asnEncodedData in attribute.Values)
+                    {
+                        existing.Values.Add(asnEncodedData);
+                    }
+                    return index;
+                }
+            }
+
+            int indexOfNewItem = _list.Count;
+            _list.Add(attribute);
+            return indexOfNewItem;
+        }
+
+        public void Remove(CryptographicAttributeObject attribute)
+        {
+            if (attribute == null)
+                throw new ArgumentNullException(nameof(attribute));
+
+            _list.Remove(attribute);
+        }
+
+        public CryptographicAttributeObject this[int index]
+        {
+            get
+            {
+                return _list[index];
+            }
+        }
+
+        public int Count
+        {
+            get
+            {
+                return _list.Count;
+            }
+        }
+
+        public CryptographicAttributeObjectEnumerator GetEnumerator()
+        {
+            return new CryptographicAttributeObjectEnumerator(this);
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return new CryptographicAttributeObjectEnumerator(this);
+        }
+
+        void ICollection.CopyTo(Array array, int index)
+        {
+            if (array == null)
+                throw new ArgumentNullException(nameof(array));
+            if (array.Rank != 1)
+                throw new ArgumentException(SR.Arg_RankMultiDimNotSupported);
+            if (index < 0 || index >= array.Length)
```

Comment:
Technically this prevents the empty collection to be copied to an empty array.  But since it already exists in desktop this way it's probably better to leave it as is.